Add initial content outline to page for Dev/Prod Modes (#103)
streamingfast/substreams: docs/developer-guide/running-substreams.md

```

The example output shown above contains data for different transfers from data in the blockchain. These transfers can also be verified on [Etherscan](https://etherscan.io/tx/0xcfb197f62ec5c7f0e71a11ec0c4a0e394a3aa41db5386e85526f86c84b3f2796).

## Development & Production Mode

### Development Mode

Production and development modes impact the execution of Substreams. Key aspects of execution include:

- The time required to reach the first byte.
- The speed that large ranges get executed.
- The module logs and outputs that are sent back to the client.

Differences between production and development modes include:

> - Forward execution is enabled in production mode and disabled in development mode.

Reviewer: Let's link `Forward execution` to where it's defined, let's also use `Forward parallel execution`

Author: Wording updated to add "parallel". Unsure where to link exactly. The changes will be available in the next commit/push.

- The time required to reach the first byte in development mode is faster than in production mode.

> - In development mode, the client will receive all of the executed module's logs. In production mode the client will only receive the output module's logs.

Reviewer: There is no logs at all in production mode, no where in the module's tree.

Author: The verbiage for production mode logs has been removed in its entirety. The changes will be available in the next commit/push.

- In development mode, it's possible to request specific store snapshots in the execution tree.

> 

Reviewer: We are also missing the fact the in development mode, multiple module's output is possible.

Author: Added a bullet point/sentence for multiple output. Please review for technical accuracy. It reads a bit strange to me as it stands. The changes will be available in the next commit/push.

Reviewer: > It's important to remember the -p flag is passed to the substreams CLI only in development mode.

Reading this is super confusing to me. Also, it should be `Substreams CLI`.

Author: Sentence removed. Per the style guide when we refer to the Substreams CLI is should be formatting as:

`substreams` CLI

So, lower case here for Substreams and use the code font.

> Examples: (given the dependencies: `[block] --> [map_pools] --> [store_pools] --> [map_transfers])`

Reviewer: This would require a quick sentence to introduce the examples, the transition feels really hard here. It seems also the bullet list below is all for development mode but no where the reader can know that.

Author: Subtitle/header added for this section. The changes will be available in the next commit/push.

Reviewer: > Examples: (given the dependencies: `[block] --> [map_pools] --> [store_pools] --> [map_transfers])`

Still feels like it should be improved. I think it should be turned into a proper sentence or 2.

> - Running the `substreams run substreams.yaml map_transfers` command will only print the `map_transfers` module's outputs and logs.

Reviewer: It's not clear in the list below if running in production mode or development mode. I would introduce that fact the development is enabled by default in the CLI unless `-p` is specified. Also, a note below that should be added to emphasis the fact that we expect people to run the production mode using a Substreams `sink`. Very very rarely people should use `-p` as `substreams` CLI is really more a development tool than a production tool.

Author: Verbiage added to address the -p flag and development mode versus production mode. The changes will be available in the next commit/push.

Reviewer: Let's add on each of them (list bullet points) that is runs in development mode.
```suggestion
- Running the `substreams run substreams.yaml map_transfers` command run in development mode and only print the `map_transfers` module's outputs and logs.
```

Something like this.

Author: Updated to reflect the content you've provided. Thank you. Changes will be available for review in the next commit/push.

- Running the `substreams run substreams.yaml map_transfers --debug-modules-output=map_pools,map_transfers,store_pools` command will print the outputs of the `map_pools`, `map_transfers`, and `store_pools` modules.
- Running the `substreams run substreams.yaml map_transfers -s 1000 -t +5 --debug-modules-initial-snapshot=store_pools` command will print all the entries in the `store_pools` module at block 999, then continue with outputs and logs from the `map_transfers` module in blocks 1000 through 1004.

> ### Parallel Execution for Substreams Modes

Reviewer: The `Substreams Modes` here feel weird a bit. What do you think about:
```suggestion
### Parallel Execution in Development vs Production Modes
```

I leave you the final say on it.

Author: Updated in the organization of new content. Please review for technical accuracy. The changes will be available in the next commit/push.

> The two steps involved during parallel execution are backward execution and forward execution.

Reviewer: Let's use everywhere `backward parallel execution` and `forward parallel execution`.

```suggestion
The two steps involved during parallel execution are backward parallel execution and forward parallel execution.
```

Author: This has been updated through the content. Please let me know if you spot any additional instances. The changes will be available in the next commit/push.

Backward execution consists of executing in parallel block ranges, from the module's initial block, up to the start block of the request. If the start block of the request matches the module's initial block no backwards execution is performed.

> Forward execution consists of executing in parallel block ranges from the start block of the request up to last known final block, also called an irreversible block, or the stop block of the request depending on which is smaller. Forward execution significantly improves the performance of Substreams, however the ability to stream module logs is lost.

Reviewer: We should not mention that fact that module's log are lost, it would be technically feasible even in forward parallel execution but we decided not. The fact that there is no logs should be only mention because of `production` vs `development`.

Author: Updated to remove the wording referencing any lost logs. The changes will be available in the next commit/push.

```

The example output shown above contains data for different transfers from data in the blockchain. These transfers can also be verified on [Etherscan](https://etherscan.io/tx/0xcfb197f62ec5c7f0e71a11ec0c4a0e394a3aa41db5386e85526f86c84b3f2796).

## Development abd production mode

### Development mode

Production and development modes impact the execution of Substreams. Key aspects of execution include:

- The time required to reach the first byte.
- The speed that large ranges get executed.
- The module logs and outputs that are sent back to the client.

## Parallel execution in development versus production modes

Differences between production and development modes include:

- Forward parallel execution is enabled in production mode and disabled in development mode.
- The time required to reach the first byte in development mode is faster than in production mode.

> Specific attributes of development modes include:

Reviewer: ```suggestion
Specific attributes of development mode include:
```

Author: Great catch! Update made. Changes available in next push.

- The client will receive all of the executed module's logs.
- It's possible to request specific store snapshots in the execution tree.
- Multiple module's output is possible.

## Production and development parallel execution examples

In most cases, you will run production mode, using a Substreams sink. Development mode is enabled by default in the CLI unless -p is specified.

Examples: (given the dependencies: `[block] --> [map_pools] --> [store_pools] --> [map_transfers])`

> - Running the `substreams run substreams.yaml map_transfers` command `run` in development mode and only print the `map_transfers` module's outputs and logs.

Reviewer: You used `run` but I meant it to be a word. My goal here is to make it clear to the user that he is running in development mode. 

```suggestion
- Running the `substreams run substreams.yaml map_transfers` command executes in development mode and only print the `map_transfers` module's outputs and logs.
```

The added part about development mode must also be added to the two other list item.

Author: Updates made and applied to all three list items. Please review for technical accuracy. Changes available in next push.
